### ISSUE TITLE
Ship normalization/transformation into the model buffers (*.pt format) rather than applying externally the dataset level

### ISSUE THREAD

**Issue by georgemilosh** [georgemilosh/closure#19]
Goal: refactor the training and deployment workflow so that feature/target transformations and normalization are no longer applied externally at the dataset level, but are encoded inside the model via registered buffers and tensor-only preprocessing/postprocessing logic. Preserve training behavior where needed, and support both Python checkpointing (.pth) and TorchScript deployment (.pt) for CUDA/C++.

Current workflow
	•	Inputs and targets are transformed outside the model, at dataset or dataloader level.
	•	Normalization statistics are stored externally and passed around separately.
	•	Model is trained on already-transformed / already-normalized tensors.
	•	This makes deployment fragile because inference requires reproducing preprocessing exactly outside the model.

Target workflow
	•	Raw physical inputs should be accepted by the deployed model.
	•	Feature transforms, target transforms, normalization, and inverse normalization should live inside the model.
	•	Transformation constants and normalization statistics should be stored as model buffers.
	•	The TorchScript-exported model should be self-contained: raw input -> preprocess -> network -> postprocess -> raw output.
	•	Python training should still support exact equivalence with the previous dataset-level workflow.

Key design principles
	1.	Put all required inference logic inside the model if inference would be wrong without it.
	2.	Use register_buffer(...) for normalization statistics and transformation constants, not plain attributes.
	3.	Avoid registering buffers as None for the required inference state. Register concrete tensors in __init__, then overwrite values later.
	4.	Use only Torch tensor operations in model preprocessing/postprocessing intended for TorchScript. No NumPy in scripted paths.
	5.	Keep forward() as the deployment inference contract.
	6.	Separate training-space outputs from deployment-space outputs if exact equivalence is required.

Recommended model split

A. Core model
	•	Pure neural network.
	•	Input: normalized/transformed features.
	•	Output: normalized/transformed targets.
	•	This reproduces the old mathematical training problem.

B. Deployment wrapper
	•	Contains:
	•	feature transforms, e.g., log, asinh, signed-log, clipping
	•	input normalization
	•	call to core model
	•	output de-normalization
	•	inverse target transforms, e.g., exp, sinh
	•	Input: raw physical state
	•	Output: raw physical prediction
	•	Export this wrapper to TorchScript .pt.

Why this split
	•	Training remains exactly equivalent to the old workflow if the loss is still computed in the transformed/normalized target space.
	•	Deployment becomes self-contained and robust.
	•	Avoids accidentally changing the optimization objective.

Important equivalence point
These are equivalent only if the loss is computed in the same space as before.

Old training:
	•	x_raw -> T_x -> normalize -> core_net -> y_norm_pred
	•	loss against y_norm_true

Equivalent new training:
	•	model preprocessing does T_x and normalization internally
	•	model returns y_norm_pred
	•	loss still uses normalized/transformed targets

Not equivalent:
	•	de-normalize and inverse-transform inside training forward
	•	compute loss in raw target space

That changes the objective and may change the learned model.

Performance note
	•	Putting log, asinh, affine normalization, exp, etc., into the graph should not materially slow training relative to the MLP itself.
	•	These elementwise ops are cheap compared with linear layers/backprop through weights.
	•	Autograd flowing through fixed preprocessing is expected and generally not a problem.

Implementation instructions
	1.	Replace dataset-level transforms

	•	Stop permanently transforming inputs/targets inside the dataset object for deployment purposes.
	•	Dataset should ideally provide raw tensors, or at least keep a path to raw tensors available.

	2.	Add explicit preprocessing/postprocessing methods

	•	preprocess_inputs(x_raw) -> x_norm
	•	postprocess_outputs(y_norm) -> y_raw
	•	Use Torch ops only.

	3.	Store constants as buffers
Examples:

	•	x_mean, x_std, y_mean, y_std
	•	per-channel scaling constants such as B_scale
	•	small eps for safe log/clamp

	4.	Register buffers once in __init__
Example pattern:

	•	self.register_buffer("x_mean", torch.zeros(n_inputs))
	•	self.register_buffer("x_std", torch.ones(n_inputs))
	•	self.register_buffer("y_mean", torch.zeros(n_outputs))
	•	self.register_buffer("y_std", torch.ones(n_outputs))

Then later populate with .copy_(...) or assignment to buffer data, not repeated register_buffer(...).
	5.	Keep NumPy helpers out of TorchScript path

	•	Python-only methods like predict_np(...) can remain as convenience wrappers, but they should call tensor-based methods and are not part of scripted deployment.

	6.	Decide on the training API
Recommended:

	•	core model returns normalized/transformed targets for training
	•	wrapper model returns raw outputs for deployment
Alternative:
	•	one model exposes a tensor-only method for training-space outputs and uses forward() for deployment-space outputs

Feature/target transformation rules
	•	Normalization statistics must be computed in transformed space, not raw space, if the network is trained on transformed features/targets.
	•	Example:
	•	if input channel is log(rho), then x_mean/x_std must be statistics of log(rho)
	•	if target is log(p_par), then y_mean/y_std must be statistics of log(p_par)

Domain safety
	•	Any log transform must guard against invalid input.
	•	Use a clear policy such as:
	•	torch.clamp(x, min=eps) before log
	•	or use a transform defined on all reals, e.g., asinh or signed-log
	•	The policy used in training must match deployment.

Checkpointing policy
	1.	Python training checkpoint: .pth
Purpose:

	•	resume training
	•	reproducibility
	•	experimentation

Should contain at least:
	•	core_model_state_dict
	•	wrapper/model state if wrapper is also trainable or needed
	•	optimizer state
	•	scheduler state
	•	training config / architecture config
	•	normalization statistics and transform metadata if not already embedded as buffers
	•	epoch / step / metrics as needed

If normalization and transform constants are already buffers in the saved model state_dict, storing them again separately is optional and mostly for readability/backward compatibility.
	2.	Deployment artifact: TorchScript .pt
Purpose:

	•	inference in Python without external preprocessing
	•	CUDA / LibTorch / C++ deployment

Should contain:
	•	self-contained wrapped model
	•	buffers for all transforms and normalization constants
	•	forward(raw_input) -> raw_output

Typical export flow
	•	populate buffers
	•	model.eval()
	•	scripted = torch.jit.script(deployment_wrapper)
	•	scripted.save("model.pt")

Do not rely on external normalization arrays or Python preprocessing code for .pt deployment.

Practical recommendation
Maintain two artifacts:
	•	.pth for research/training/resume
	•	.pt for deployment

This is the intended split:
	•	.pth: rich checkpoint, may include optimizer/scheduler/etc.
	•	.pt: minimal, self-contained inference model

Concrete behavioral contract to implement
Training contract:
	•	The raw dataset can be converted to transformed/normalized targets consistently
	•	loss remains in the same transformed/normalized space as the previous workflow unless intentionally changed

Deployment contract:
	•	user passes raw state tensor shaped (..., n_inputs)
	•	model performs internal transform + normalize + inference + inverse normalize + inverse target transform
	•	output is raw physical prediction shaped (..., n_outputs)

Example schema for current plasma surrogate
Inputs:
	•	(rho, ux, uy, uz, Bx, By, Bz)

Possible transforms:
	•	rho -> log(rho)
	•	u -> unchanged
	•	B -> asinh(B / B_scale) or asinh(B)

Targets:
	•	network trained on log(p_par), log(p_perp)
	•	postprocess returns exp(...) to recover physical pressures

Requested refactor outcome
	•	remove fragile external dependence on dataset-level normalization during inference
	•	make model serialization self-sufficient
	•	preserve old training behavior
	•	support clean TorchScript export for future CUDA/C++ use

Acceptance criteria
	1.	A saved .pth can restore training exactly.
	2.	A saved .pt can run on raw inputs without any external normalization arrays.
	3.	Moving the model to CPU/GPU moves normalization buffers automatically.
	4.	Scripted model contains no NumPy-dependent inference code.
	5.	Predictions from old external-preprocessing pipeline and new buffer-based pipeline match to numerical tolerance when evaluated in equivalent mode.
:::

A shorter version:

“Refactor the model pipeline so that all feature transforms and normalization are stored inside the model as registered buffers and tensor-only preprocessing/postprocessing methods. Preserve exact training behavior by keeping the loss in transformed/normalized target space. Create a deployment wrapper whose forward() accepts raw physical inputs and returns raw physical outputs. Save .pth checkpoints for training/resume and export a self-contained TorchScript .pt for inference in CUDA/C++.”

In most cases, updating already trained models does not require retraining. What you usually do is wrap the trained core network in a new deployment module that reproduces the old preprocessing and postprocessing steps internally.

Conceptually:

old setup:
raw input -> external transform/normalize -> old model -> external inverse steps

new setup:
raw input -> new wrapper does transform/normalize -> old model -> new wrapper does inverse steps

So if the old model was trained on normalized/transformed tensors, you can often keep its weights exactly as they are and just embed the old preprocessing into a wrapper.

There are really three levels of migration difficulty.

Easiest case:
you have the old state_dict, architecture, normalization arrays, and clear knowledge of feature/target transforms.
Then migration is straightforward. You instantiate the old model, load weights, create a wrapper with the saved transforms/stats as buffers, and export that wrapper. Numerically, it should reproduce the old inference path up to tolerance.

Moderate case:
you have the weights, but normalization or transforms were not saved cleanly and are scattered in training scripts or dataset code.
Then migration is still possible, but you need some archaeology: recover exact means/stds, transform conventions, channel ordering, epsilon/clipping policy, whether targets were logged before normalization or vice versa, and so on. This is usually the real source of pain, not the wrapper itself.

Hard case:
you no longer know the exact preprocessing used during training.
Then exact migration may be impossible without reconstructing it from code, old checkpoints, cached datasets, or validation behavior. Even small ambiguities can matter, such as:
	•	was rho transformed by log(rho) or log10(rho)?
	•	was B transformed by asinh(B) or asinh(B/B0)?
	•	were stds computed before or after filtering out bad points?
	•	were targets normalized after log transform or before?

If those are uncertain, the wrapped model may not reproduce the old one exactly.

Option 1: keep the old core unchanged and add a wrapper around it.
This is the safest and recommended migration path.

Option 2: fold some affine normalization into the first and last linear layers.
This is mathematically possible for purely affine preprocessing/postprocessing, but once you have nonlinear transforms like log or asinh, those cannot be folded into the linear weights. So in your case, a wrapper is the natural approach.

Option 3: fine-tune under the new architecture.
Only needed if you intentionally change the training objective or want the model itself to learn under the new integrated setup.
